runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,218 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
+ *
+ *  \details
+ *     This funcition generates a sequence to check per site cache for object class and cast class before calling out to jitInstanceOf helper
+ */
+static
+void genInstanceOfDynamicCacheAndHelperCall(TR::Node *node, TR::CodeGenerator *cg, TR::Register *castClassReg, TR::Register *objClassReg, TR::Register *resultReg, TR_S390ScratchRegisterManager *srm, TR::LabelSymbol *doneLabel, TR::LabelSymbol *helperCallLabel, TR::LabelSymbol *dynamicCacheTestLabel, TR::LabelSymbol *branchLabel, TR::LabelSymbol *trueLabel, TR::LabelSymbol *falseLabel, bool dynamicCastClass, bool generateDynamicCache, bool cacheCastClass, bool ifInstanceOf, bool trueFallThrough )
+   {
+   TR::Compilation                *comp = cg->comp();
+   bool needResult = resultReg != NULL;
+   if (!castClassReg)
+      castClassReg = cg->evaluate(node->getSecondChild());
+   int32_t maxOnsiteCacheSlots = comp->getOptions()->getMaxOnsiteCacheSlotForInstanceOf();
+   TR::Register *dynamicCacheReg = NULL;
+   int32_t addressSize = TR::Compiler->om.sizeofReferenceAddress();
+   /* Layout of the writable data snippet
+    * Case - 1 : Cast class is runtime variable
+    * [UpdateIndex][ObjClassSlot-0][CastClassSlot-0]...[ObjClassSlot-N][CastClassSlot-N]
+    * Case - 2 : Cast Class is interface / unresolved
+    * [UpdateIndex][ObjClassSlot-0]...[ObjClassSlot-N]
+    * If there is only one cache slot, we will not have header.
+    * Last bit of cached objectClass will set to 1 indicating false cast
+    */
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   if (generateDynamicCache)
+      {
+      TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
+      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
+       * Case 1 : Cast class is runtime variable
+       * Case 2 : Cast class is interface / unresolved
+       * Number Of Slots |  Bytes needed for Case 1 | Bytes needed for Case 2
+       *        1        |              16          |           8
+       *        2        |              64          |           32
+       *        3        |              64          |           32
+       *        4        |              128         |           64
+       *        5        |              128         |           64
+       *        6        |              128         |           64
+       */
+      int32_t requestedBytes = 1 << (int) (log2(snippetSizeInBytes-1)+1);
+      traceMsg(comp, "Requested Bytes = %d\n",requestedBytes);
+      // NOTE: For single slot cache, we initialize snippet with addressSize (4/8) assuming which can not be objectClass
+      // In all cases, we use first addressSize bytes to store offset of the circular buffer and rest of buffer will be initialized with 0xf.
+      TR_ASSERT(maxOnsiteCacheSlots <= 7, "InstanceOf : Maximum 7 slots per site allowed");
```

Comment:
This should be a fatal assert. And we should improve the message as to why the maximum slots is 7. Perhaps something like this:

```
TR_ASSERT_FATAL(maxOnsiteCacheSlots <= 7, "Maximum 7 slots per site allowed because we use a fixed stack allocated buffer to construct the snippet\n");
```